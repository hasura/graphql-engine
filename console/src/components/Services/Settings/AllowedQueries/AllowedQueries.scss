@import '../Settings.scss';

.ul_left_small {
  -webkit-padding-start: 15px;
}

.subsection {
  padding-left: 15px;
  padding-right: 15px;
}
