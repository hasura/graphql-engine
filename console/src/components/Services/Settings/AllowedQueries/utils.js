import { parse, print, visit } from 'graphql';

export const readFile = (file, callback) => {
  const reader = new FileReader();
  reader.onload = event => {
    const content = event.target.result;
    callback(content);
  };

  reader.onerror = event => {
    console.error('File could not be read! Code ' + event.target.error.code);
  };

  reader.readAsText(file);
};

function recurQueryDef(queryDef, fragments, definitionHash) {
  visit(queryDef, {
    FragmentSpread(node) {
      fragments.add(node.name.value);
      recurQueryDef(definitionHash[node.name.value], fragments, definitionHash);
    },
  });
}

const getQueryFragments = (queryDef, definitionHash = {}) => {
  const fragments = new Set();
  recurQueryDef(queryDef, fragments, definitionHash);
  return [...fragments];
};

const getQueryString = (queryDef, fragmentDefs, definitionHash = {}) => {
  let queryString = print(queryDef);

  const queryFragments = getQueryFragments(queryDef, definitionHash);

  queryFragments.forEach(qf => {
    const fragmentDef = fragmentDefs.find(fd => fd.name.value === qf);

    if (fragmentDef) {
      queryString += '\n\n' + print(fragmentDef);
    }
  });

  return queryString;
};

export const parseQueryString = queryString => {
  const queries = [];

  let parsedQueryString;

  try {
    parsedQueryString = parse(queryString);
  } catch (ex) {
    throw new Error('Parsing operation failed');
  }

  const definitionHash = (parsedQueryString.definitions || []).reduce(
    (defObj, queryObj) => {
      defObj[queryObj.name.value] = queryObj;
      return defObj;
    },
    {}
  );

  const queryDefs = parsedQueryString.definitions.filter(
    def => def.kind === 'OperationDefinition'
  );

  const fragmentDefs = parsedQueryString.definitions.filter(
    def => def.kind === 'FragmentDefinition'
  );

  queryDefs.forEach(queryDef => {
    if (!queryDef.name) {
      throw new Error(`Operation without name found: ${print(queryDef)}`);
    }

    const query = {
      name: queryDef.name.value,
      query: getQueryString(queryDef, fragmentDefs, definitionHash),
    };

    queries.push(query);
  });

  const queryNames = queries.map(q => q.name);
  const duplicateNames = queryNames.filter(
    (q, i) => queryNames.indexOf(q) !== i
  );
  if (duplicateNames.length > 0) {
    throw new Error(
      `Operations with duplicate names found: ${duplicateNames.join(', ')}`
    );
  }

  return queries;
};
