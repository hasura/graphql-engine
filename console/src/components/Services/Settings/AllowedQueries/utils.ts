import { parse, print, visit, DefinitionNode } from 'graphql';
import { AllowedQueriesCollection } from '../../../../metadata/reducer';

export type NewDefinitionNode = DefinitionNode & {
  name?: {
    value: string;
  };
};

export const readFile = (
  file: File | null,
  callback: (content: string) => void
) => {
  const reader = new FileReader();
  reader.onload = event => {
    const content = event.target!.result as string;
    callback(content);
  };

  reader.onerror = event => {
    console.error(`File could not be read! Code ${event.target!.error!.code}`);
  };

  if (file) reader.readAsText(file);
};

const recurQueryDef = (
  queryDef: NewDefinitionNode,
  fragments: Set<string>,
  definitionHash: Record<string, any>
) => {
  visit(queryDef, {
    FragmentSpread(node) {
      fragments.add(node.name.value);
      recurQueryDef(definitionHash[node.name.value], fragments, definitionHash);
    },
  });
};

const getQueryFragments = (
  queryDef: NewDefinitionNode,
  definitionHash: Record<string, any> = {}
) => {
  const fragments = new Set<string>();
  recurQueryDef(queryDef, fragments, definitionHash);
  return [...Array.from(fragments)];
};

const getQueryString = (
  queryDef: NewDefinitionNode,
  fragmentDefs: NewDefinitionNode[],
  definitionHash: Record<string, any> = {}
) => {
  let queryString = print(queryDef);

  const queryFragments = getQueryFragments(queryDef, definitionHash);

  queryFragments.forEach(qf => {
    // eslint-disable-next-line array-callback-return
    const fragmentDef = fragmentDefs.find(fd => {
      if (fd.name) return fd.name.value === qf;
    });

    if (fragmentDef) {
      queryString += `\n\n${print(fragmentDef)}`;
    }
  });

  return queryString;
};

export const parseQueryString = (queryString: string) => {
  const queries: { name: string; query: string }[] = [];

  let parsedQueryString;

  try {
    parsedQueryString = parse(queryString);
  } catch (ex) {
    throw new Error('Parsing operation failed');
  }

  const definitions: NewDefinitionNode[] = [...parsedQueryString.definitions];

  const definitionHash = (definitions || []).reduce(
    (defObj: Record<string, NewDefinitionNode>, queryObj) => {
      if (queryObj.name) defObj[queryObj.name.value] = queryObj;
      return defObj;
    },
    {}
  );

  const queryDefs = definitions.filter(
    def => def.kind === 'OperationDefinition'
  );

  const fragmentDefs = definitions.filter(
    def => def.kind === 'FragmentDefinition'
  );

  queryDefs.forEach(queryDef => {
    if (!queryDef.name) {
      throw new Error(`Operation without name found: ${print(queryDef)}`);
    }

    const query = {
      name: queryDef.name.value,
      query: getQueryString(queryDef, fragmentDefs, definitionHash),
    };

    queries.push(query);
  });

  const queryNames = queries.map(q => q.name);
  const duplicateNames = queryNames.filter(
    (q, i) => queryNames.indexOf(q) !== i
  );
  if (duplicateNames.length > 0) {
    throw new Error(
      `Operations with duplicate names found: ${duplicateNames.join(', ')}`
    );
  }

  return queries;
};

export const getQueriesInCollection = (
  collectionName: string,
  allowedQueries: AllowedQueriesCollection[]
) => {
  const queries: AllowedQueriesCollection[] = [];
  allowedQueries.forEach(query => {
    if (query.collection === collectionName) {
      queries.push(query);
    }
  });
  return queries;
};

export const checkLastQuery = (
  collectionName: string,
  queries: AllowedQueriesCollection[]
) => {
  return getQueriesInCollection(collectionName, queries).length === 1;
};

// Missing feature in typescript https://stackoverflow.com/questions/33464504/using-spread-syntax-and-new-set-with-typescript/33464709
export const getCollectionNames = (queries: AllowedQueriesCollection[]) => {
  return Array.from(new Set(queries.map(query => query.collection)));
};
