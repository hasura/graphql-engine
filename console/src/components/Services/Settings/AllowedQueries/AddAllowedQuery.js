import React from 'react';
import AceEditor from 'react-ace';
import styles from './AllowedQueries.scss';

import ExpandableEditor from '../../../Common/Layout/ExpandableEditor/Editor';
import Tooltip from '../../../Common/Tooltip/Tooltip';

import { readFile, parseQueryString } from './utils';
import { showErrorNotification } from '../../Common/Notification';
import { addAllowedQueries } from '../../../../metadata/actions';

class AddAllowedQuery extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      manualQuery: {},
      graphqlFile: null,
    };
  }

  render() {
    const { dispatch, isEmptyList } = this.props;
    const { manualQuery, graphqlFile } = this.state;

    const handleManualCollapse = () => {
      this.setState({ manualQuery: {} });
    };

    const handleManualSubmit = toggle => {
      dispatch(addAllowedQueries([manualQuery], isEmptyList, toggle));
    };

    const handleFileUploadCollapse = () => {};

    function handleFileUploadSubmit(toggle) {
      const addFileQueries = content => {
        try {
          const fileQueries = parseQueryString(content);
          dispatch(addAllowedQueries(fileQueries, isEmptyList, toggle));
        } catch (error) {
          dispatch(
            showErrorNotification('Uploading operations failed', error.message)
          );
        }
      };

      readFile(graphqlFile, addFileQueries);
    }

    const getManualQueryInput = () => {
      const getNameInput = () => {
        const handleNameChange = e => {
          this.setState({
            manualQuery: {
              ...manualQuery,
              name: e.target.value,
            },
          });
        };

        return (
          <div>
            <div className={styles.add_mar_bottom_mid}>
              <b>Operation name:</b>
            </div>
            <input
              type="text"
              className={'form-control input-sm ' + styles.inline_block}
              placeholder={'operation_name'}
              value={manualQuery.name}
              onChange={handleNameChange}
            />
          </div>
        );
      };

      const getQueryInput = () => {
        const handleQueryChange = val => {
          this.setState({
            manualQuery: {
              ...manualQuery,
              query: val,
            },
          });
        };

        return (
          <div>
            <div className={styles.add_mar_bottom_mid}>
              <b>Operation:</b>
            </div>
            <AceEditor
              data-test="allowed_operation_add"
              mode="graphql"
              theme="github"
              name="allowed_operation_add"
              value={manualQuery.query}
              minLines={8}
              maxLines={100}
              width="100%"
              showPrintMargin={false}
              onChange={handleQueryChange}
            />
          </div>
        );
      };

      return (
        <div>
          <div>{getNameInput()}</div>
          <div className={styles.add_mar_top}>{getQueryInput()}</div>
        </div>
      );
    };

    const getFileUploadInput = () => {
      const handleFileUpload = e => {
        const files = e.target.files;
        this.setState({ graphqlFile: files[0] });
      };

      return (
        <div>
          <div className={styles.add_mar_bottom_mid}>
            <b>Graphql File:</b>
            <Tooltip message={'.graphql file with operations'} />
          </div>
          <input
            type="file"
            className={'form-control input-sm ' + styles.inline_block}
            onChange={handleFileUpload}
          />
        </div>
      );
    };

    return (
      <div>
        <h4 className={styles.subheading_text}>
          Add new operations to allow-list
        </h4>
        <div className={styles.subsection}>
          <div>
            <ExpandableEditor
              expandButtonText="Add operation manually"
              editorExpanded={getManualQueryInput}
              collapseCallback={handleManualCollapse}
              property="add-allowed-operation"
              service="add-allowed-operation"
              saveButtonText="Add"
              saveFunc={handleManualSubmit}
            />
          </div>
          <div className={styles.add_mar_top}>OR</div>
          <div className={styles.add_mar_top}>
            <ExpandableEditor
              expandButtonText="Upload graphql file"
              editorExpanded={getFileUploadInput}
              collapseCallback={handleFileUploadCollapse}
              property="upload-allowed-operations"
              service="upload-allowed-operations"
              saveButtonText="Upload"
              saveFunc={handleFileUploadSubmit}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default AddAllowedQuery;
