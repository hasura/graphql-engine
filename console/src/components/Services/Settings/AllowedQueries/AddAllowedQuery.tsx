import React, { useState } from 'react';
import AceEditor from 'react-ace';
import styles from './AllowedQueries.scss';

import ExpandableEditor from '../../../Common/Layout/ExpandableEditor/Editor';
import Tooltip from '../../../Common/Tooltip/Tooltip';

import { readFile, parseQueryString } from './utils';
import { showErrorNotification } from '../../Common/Notification';
import { addAllowedQueries } from '../../../../metadata/actions';
import { allowedQueriesCollection } from '../../../../metadata/utils';
import { AllowedQueriesCollection } from '../../../../metadata/reducer';
import { Dispatch } from '../../../../types';

const defaultManualQuery: AllowedQueriesCollection = {
  name: '',
  query: '',
  collection: allowedQueriesCollection,
};

type AddAllowedQueryProps = {
  dispatch: Dispatch;
};

const AddAllowedQuery: React.FC<AddAllowedQueryProps> = props => {
  const { dispatch } = props;

  const [manualQuery, setManualQuery] = useState<AllowedQueriesCollection>(
    defaultManualQuery
  );
  const [graphqlFile, setGraphqlFile] = useState<File | null>(null);

  const handleManualCollapse = () => {
    setManualQuery(defaultManualQuery);
  };

  const handleManualSubmit = (toggle: () => void) => {
    dispatch(addAllowedQueries([manualQuery], toggle));
  };

  const handleFileUploadCollapse = () => {};

  const handleFileUploadSubmit = (toggle: () => void) => {
    const addFileQueries = (content: string) => {
      try {
        const fileQueries = parseQueryString(content);
        dispatch(addAllowedQueries(fileQueries, toggle));
      } catch (error) {
        dispatch(
          showErrorNotification('Uploading operations failed', error.message)
        );
      }
    };

    readFile(graphqlFile, addFileQueries);
  };

  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setManualQuery({
      ...manualQuery,
      name: e.target.value,
    });
  };

  const handleQueryChange = (val: string) => {
    setManualQuery({
      ...manualQuery,
      query: val,
    });
  };

  const manualQueryInput = () => (
    <div>
      <div>
        <div className={styles.add_mar_bottom_mid}>
          <b>Operation name:</b>
        </div>
        <input
          type="text"
          className={`form-control input-sm ${styles.inline_block}`}
          placeholder="operation_name"
          value={manualQuery.name}
          onChange={handleNameChange}
        />
      </div>
      <div className={styles.add_mar_top}>
        <div>
          <div className={styles.add_mar_bottom_mid}>
            <b>Operation:</b>
          </div>
          <AceEditor
            data-test="allowed_operation_add"
            mode="graphql"
            theme="github"
            name="allowed_operation_add"
            value={manualQuery.query}
            minLines={8}
            maxLines={100}
            width="100%"
            showPrintMargin={false}
            onChange={handleQueryChange}
          />
        </div>
      </div>
    </div>
  );

  const handleFileUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files;
    setGraphqlFile(files![0]);
  };

  const fileUploadInput = () => (
    <div>
      <div className={styles.add_mar_bottom_mid}>
        <b>Graphql File:</b>
        <Tooltip message=".graphql file with operations" />
      </div>
      <input
        type="file"
        className={`form-control input-sm ${styles.inline_block}`}
        onChange={handleFileUpload}
      />
    </div>
  );

  return (
    <div>
      <h4 className={styles.subheading_text}>
        Add new operations to allow-list
      </h4>
      <div className={styles.subsection}>
        <div>
          <ExpandableEditor
            expandButtonText="Add operation manually"
            editorExpanded={manualQueryInput}
            collapseCallback={handleManualCollapse}
            property="add-allowed-operation"
            service="add-allowed-operation"
            saveButtonText="Add"
            saveFunc={handleManualSubmit}
          />
        </div>
        <div className={styles.add_mar_top}>OR</div>
        <div className={styles.add_mar_top}>
          <ExpandableEditor
            expandButtonText="Upload graphql file"
            editorExpanded={fileUploadInput}
            collapseCallback={handleFileUploadCollapse}
            property="upload-allowed-operations"
            service="upload-allowed-operations"
            saveButtonText="Upload"
            saveFunc={handleFileUploadSubmit}
          />
        </div>
      </div>
    </div>
  );
};

export default AddAllowedQuery;
