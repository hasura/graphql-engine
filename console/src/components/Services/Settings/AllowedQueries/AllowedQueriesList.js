import React from 'react';
import AceEditor from 'react-ace';

import ExpandableEditor from '../../../Common/Layout/ExpandableEditor/Editor';
import styles from './AllowedQueries.scss';
import Button from '../../../Common/Button/Button';

import { getConfirmation } from '../../../Common/utils/jsUtils';
import {
  updateAllowedQuery,
  deleteAllowedQuery,
  deleteAllowList,
} from '../../../../metadata/actions';

class AllowedQueriesList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modifiedQueries: {},
    };
  }

  render() {
    const { allowedQueries, dispatch } = this.props;
    const { modifiedQueries } = this.state;

    const getQueryList = () => {
      if (allowedQueries.length === 0) {
        return <div>No operations in allow-list yet</div>;
      }

      return allowedQueries.map((query, i) => {
        const queryName = query.name;

        const collapsedLabel = () => (
          <div>
            <b>{queryName}</b>
          </div>
        );

        const expandedLabel = collapsedLabel;

        const queryEditorExpanded = () => {
          const modifiedQuery = modifiedQueries[queryName] || { ...query };

          const handleNameChange = e => {
            const newModifiedQueries = { ...modifiedQueries };
            newModifiedQueries[queryName].name = e.target.value;

            this.setState({ modifiedQueries: newModifiedQueries });
          };

          const handleQueryChange = val => {
            const newModifiedQueries = { ...modifiedQueries };
            newModifiedQueries[queryName].query = val;

            this.setState({ modifiedQueries: newModifiedQueries });
          };

          return (
            <div>
              <div>
                <div className={styles.add_mar_bottom_mid}>
                  <b>Operation name:</b>
                </div>
                <input
                  type="text"
                  className={'form-control input-sm ' + styles.inline_block}
                  value={modifiedQuery.name}
                  placeholder={'operation_name'}
                  onChange={handleNameChange}
                />
              </div>
              <div className={styles.add_mar_top}>
                <div className={styles.add_mar_bottom_mid}>
                  <b>Operation:</b>
                </div>
                <AceEditor
                  data-test="allowed_operation_editor"
                  mode="graphql"
                  theme="github"
                  name="allowed_operation_editor"
                  value={modifiedQuery.query}
                  minLines={8}
                  maxLines={100}
                  width="100%"
                  showPrintMargin={false}
                  onChange={handleQueryChange}
                />
              </div>
            </div>
          );
        };

        const editorExpandCallback = () => {
          const newModifiedQueries = { ...modifiedQueries };
          newModifiedQueries[queryName] = { ...query };

          this.setState({ modifiedQueries: newModifiedQueries });
        };

        const editorCollapseCallback = () => {
          const newModifiedQueries = { ...modifiedQueries };
          delete newModifiedQueries[queryName];

          this.setState({ modifiedQueries: newModifiedQueries });
        };

        const onSubmit = () => {
          dispatch(updateAllowedQuery(queryName, modifiedQueries[queryName]));
        };

        const onDelete = () => {
          const confirmMessage = `This will delete the operation "${queryName}" from the allow-list`;
          const isOk = getConfirmation(confirmMessage);
          if (isOk) {
            const isLastQuery = allowedQueries.length === 1;

            dispatch(deleteAllowedQuery(queryName, isLastQuery));
          }
        };

        return (
          <div key={queryName}>
            <ExpandableEditor
              editorExpanded={queryEditorExpanded}
              property={`query-${i}`}
              service="modify-allowed-operation"
              saveFunc={onSubmit}
              removeFunc={onDelete}
              collapsedClass={styles.display_flex}
              expandedLabel={expandedLabel}
              collapsedLabel={collapsedLabel}
              expandCallback={editorExpandCallback}
              collapseCallback={editorCollapseCallback}
            />
          </div>
        );
      });
    };

    const getDeleteAllBtn = () => {
      const handleDeleteAll = () => {
        const confirmMessage =
          'This will delete all operations from the allow-list';
        const isOk = getConfirmation(confirmMessage, true);
        if (isOk) {
          dispatch(deleteAllowList());
        }
      };

      return (
        <Button
          size="xs"
          onClick={handleDeleteAll}
          disabled={allowedQueries.length === 0}
        >
          Delete all
        </Button>
      );
    };

    return (
      <div>
        <h4 className={styles.subheading_text}>
          Allow List
          <span className={styles.add_mar_left}>{getDeleteAllBtn()}</span>
        </h4>

        <div className={styles.subsection}>{getQueryList()}</div>
      </div>
    );
  }
}

export default AllowedQueriesList;
