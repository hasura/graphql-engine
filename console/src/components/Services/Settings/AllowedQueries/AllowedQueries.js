import React from 'react';

import AllowedQueriesNotes from './AllowedQueriesNotes';
import AddAllowedQuery from './AddAllowedQuery';
import AllowedQueriesList from './AllowedQueriesList';

import styles from './AllowedQueries.scss';
import { getAllowedQueries } from '../../../../metadata/selector';

class AllowedQueries extends React.Component {
  render() {
    const { dispatch, allowedQueries } = this.props;

    return (
      <div
        className={`${styles.clear_fix} ${styles.padd_left} ${styles.padd_top} ${styles.metadata_wrapper} container-fluid`}
      >
        <div className={styles.subHeader}>
          <h2 className={styles.headerText}>Allow List</h2>
          <div className={styles.add_mar_top + ' ' + styles.wd60}>
            <AllowedQueriesNotes />
            <hr />
            <AddAllowedQuery
              dispatch={dispatch}
              isEmptyList={allowedQueries.length === 0}
            />
            <hr />
            <AllowedQueriesList
              dispatch={dispatch}
              allowedQueries={allowedQueries}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    allowedQueries: getAllowedQueries(state),
  };
};

const allowedQueriesConnector = connect =>
  connect(mapStateToProps)(AllowedQueries);

export default allowedQueriesConnector;
