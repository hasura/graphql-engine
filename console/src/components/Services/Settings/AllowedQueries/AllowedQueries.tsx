import React from 'react';

import AllowedQueriesNotes from './AllowedQueriesNotes';
import AddAllowedQuery from './AddAllowedQuery';
import AllowedQueriesList from './AllowedQueriesList';

import styles from './AllowedQueries.scss';
import { getAllowedQueries } from '../../../../metadata/selector';
import { Dispatch, ReduxState } from '../../../../types';
import { mapDispatchToPropsEmpty } from '../../../Common/utils/reactUtils';
import { AllowedQueriesCollection } from '../../../../metadata/reducer';

interface Props {
  dispatch: Dispatch;
  allowedQueries: AllowedQueriesCollection[];
}

const AllowedQueries: React.FC<Props> = props => {
  const { dispatch, allowedQueries } = props;

  return (
    <div
      className={`${styles.clear_fix} ${styles.padd_left} ${styles.padd_top} ${styles.metadata_wrapper} container-fluid`}
    >
      <div className={styles.subHeader}>
        <h2 className={styles.headerText}>Allow List</h2>
        <div className={`${styles.add_mar_top} ${styles.wd60}`}>
          <AllowedQueriesNotes />
          <hr />
          <AddAllowedQuery dispatch={dispatch} />
          <hr />
          <AllowedQueriesList
            dispatch={dispatch}
            allowedQueries={allowedQueries}
          />
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: ReduxState) => {
  return {
    allowedQueries: getAllowedQueries(state),
  };
};

const allowedQueriesConnector = (connect: any) =>
  connect(mapStateToProps, mapDispatchToPropsEmpty)(AllowedQueries);

export default allowedQueriesConnector;
