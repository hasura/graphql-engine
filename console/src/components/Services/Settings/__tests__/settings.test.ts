import { allowedQueriesCollection } from '../../../../metadata/utils';
import { renameDuplicates, parseQueryString } from '../AllowedQueries/utils';
import { uploadedFileData } from './fixtures/allow-list';

describe('AllowedQueries_Utils.ts', () => {
  describe('renameDuplicates', () => {
    it('should rename duplicate queries', () => {
      const allowedListQueries = [
        {
          name: 'getAuthors',
          query: 'query getAuthors { author {id name} }',
          collection: allowedQueriesCollection,
        },
      ];
      const fileQueries = parseQueryString(uploadedFileData);
      const updatedQueries = renameDuplicates(fileQueries, allowedListQueries);
      expect(updatedQueries).toMatchSnapshot();
    });
  });
});
