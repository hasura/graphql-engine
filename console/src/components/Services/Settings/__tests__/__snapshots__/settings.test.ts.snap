// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllowedQueries_Utils.ts renameDuplicates should rename duplicate queries 1`] = `
Array [
  Object {
    "name": "getAuthors_1",
    "query": "query getAuthors {
  author {
    id
    name
  }
}",
  },
  Object {
    "name": "getAuthors_2",
    "query": "query getAuthors {
  author {
    id
    name
    address
  }
}",
  },
  Object {
    "name": "getAuthors_3",
    "query": "query getAuthors {
  author {
    id
    name
    age
  }
}",
  },
  Object {
    "name": "unnamed",
    "query": "{
  student {
    id
    name
    age
  }
}",
  },
  Object {
    "name": "unnamed_1",
    "query": "{
  student {
    id
    name
    roll
  }
}",
  },
  Object {
    "name": "unnamed_2",
    "query": "{
  student {
    id
    name
    address
  }
}",
  },
  Object {
    "name": "getArticles",
    "query": "query getArticles {
  article {
    id
    title
  }
}",
  },
  Object {
    "name": "getArticle",
    "query": "query getArticle {
  article {
    id
    title
    ...frag
  }
}

fragment frag on Starship {
  name
}",
  },
  Object {
    "name": "addArticles",
    "query": "mutation addArticles {
  insert_articles {
    id
    title
  }
}",
  },
]
`;
